lib/matplotlib/figure.pyi

```diff
@@ -246,7 +246,8 @@ class FigureBase(Artist):
     ) -> dict[Hashable, Axes]: ...
 
 class SubFigure(FigureBase):
-    figure: Figure
+    @property
+    def figure(self) -> Figure: ...
```

Comment:
@ksunden can you advise what I should do here?  It is a fact of the implementation that `(Sub)Figure.figure` can only ever be a `Figure`.  It is also a fact that the setter can at best be a no-op, so it doesn't really make sense to me to have a type hint encouraging the user to set it.  I can add the setter, but MyPy still complains about `Figure` being stricter than the parent's `Union[Figure, SubFigure]`.  Adding this to the allowlist doesn't seem to make any difference.  I'm stumped.

Reply:
Firstly, the allow list is only for `stubtest`, and this is breaking a `mypy` check, so that is why that is not helping. A `# type: ignore` comment could be added, but that is a bit more of a last resort. (stubtest requires mypy to pass and runs it first, part of why it takes so long)

Secondly, is there a reason to be more strict here? Subfigures can be nested inside of subfigures, can they not? Type narrowing for a subclass isn't actually the problem here, but also not clear to me why it _is_ narrowed here. (Not fully sure why you get two errors for each... but both are resolved below without adjusting their own type hints...)


Unfortunately, getting rid of the setter is breaking a pretty fundamental Object Orientation principle that all things that are provided by the parent class can be done on instances of child classes... and the method still exists, so I think type hinting it is fine...

That said, I would not be too opposed to making `Artist.figure` a read-only (type hinted at least) attribute, since I think writing to it (without using `Artist.set_figure`) is generally a mistake).

If I comment out the two lines from `artist.pyi`, I get no mypy issues found.

```diff
diff --git a/lib/matplotlib/artist.pyi b/lib/matplotlib/artist.pyi
index 08a99c0ed1..730c530dd2 100644
--- a/lib/matplotlib/artist.pyi
+++ b/lib/matplotlib/artist.pyi
@@ -33,8 +33,8 @@ class Artist:
     stale_callback: Callable[[Artist, bool], None] | None
     @property
     def figure(self) -> Figure | SubFigure: ...
-    @figure.setter
-    def figure(self, fig: Figure | SubFigure): ...
+    #@figure.setter
+    #def figure(self, fig: Figure | SubFigure): ...
     clipbox: BboxBase | None
     def __init__(self) -> None: ...
     def remove(self) -> None: ...
```


In implementation, `Artist.figure` is a standard instance attribute, made in `__init__`, so it _is_ writable by anyone... but I'm good with making the type hint stricter than that (even without making the implementation enforce it)

Also, looking back... the type hint is technically wrong, as `artist.figure` can in fact be `None`... but that may just be a "useful fib", since 99% of artists that you have will in fact have a figure, and so checking for the none case in all usages is perhaps not something we wish to enforce with type hints...

Reply:
Re type narrowing, I guess actually closely reading the code, you _do_ narrow, so not opposed to encapsulating that meaning if that is what the code does... but do have some slight hesitations there, particularly if standard artists don't get root for `art.figure`... I'd have leaned towards not special casing `Figure` there...

Reply:
> is there a reason to be more strict here? Subfigures can be nested inside of subfigures, can they not?

Unfortunately we have a mismatch in what this property is: `Artist.figure` is the parent (Sub)Figure but `FigureBase.figure` is the root `Figure`.  See https://github.com/matplotlib/matplotlib/issues/28170#issuecomment-2094524856 for why it's difficult to change that.

Reply:
https://github.com/matplotlib/matplotlib/issues/28170#issuecomment-2094722976 suggested discouraging use of the `figure` attribute.  Would it therefore be reasonable to just not type hint it?  Looks like that would require a few updates in `pyplot` to use `get_figure` instead, but that's simple enough.

Reply:
The motivation for moving from `figure` to `get_figure()` is clarity - do we want the parent or the root? Internally, we should not be using `figure` anymore. For now, it should only stay available for backward compatibility. Whether or not we want to deprecate it and force downstream users to migrate can be decided later.

I'm not clear what the effect of not type-hinting is. Would it not show up in completions? Would there be warnings on incomplete typing?

Reply:
I'm (clearly!) no expert on type hinting but I took out the hint for this property in my branch and the mypy-precommit flagged failures in `pyplot`.

> Internally, we should not be using `figure` anymore.

Should internal usage be removed in this PR or could it wait for a follow-up?  I lean towards doing it separately as I think the current change stands by itself and should be useful for the `subfigure_mosaic` work (https://github.com/matplotlib/matplotlib/issues/28170#issuecomment-2095144072).  I have started having a go at removing the internal use but there is, um, rather a lot of it and not every instance is obvious to me whether it should be root or not.  So I think separating those concerns off into a separate PR may make the review easier.

Reply:
Sure, the internal refactoring can be a follow-up.

> and not every instance is obvious to me whether it should be root or not

That's part of the motivation to get rid of it 😏. I also won't guarantee that the current usages are always correct. Subfigures (and in particular nested subfigures) are fairly recent and thus not that common.